File: 0227-basic-calculator-ii/0227-basic-calculator-ii.dart
class Solution {
  int calculate(String s) {

    List<int> stack = [];
    int num = 0;
    String op = '+';

    for (int i = 0; i <= s.length; i++) {

      String ch = (i == s.length) ? '+' : s[i];

      if (i < s.length &&
          ch.codeUnitAt(0) >= '0'.codeUnitAt(0) &&
          ch.codeUnitAt(0) <= '9'.codeUnitAt(0)) {

        num = num * 10 + int.parse(ch);
      }

      if (i == s.length ||
          ch == '+' ||
          ch == '-' ||
          ch == '*' ||
          ch == '/') {

        if (op == '+') {
          stack.add(num);
        } else if (op == '-') {
          stack.add(-num);
        } else if (op == '*') {
          stack.add(stack.removeLast() * num);
        } else if (op == '/') {
          stack.add((stack.removeLast() / num).truncate());
        }

        op = ch;
        num = 0;
      }
    }

    int ans = 0;
    for (int x in stack) {
      ans += x;
    }

    return ans;
  }
}

File: 0227-basic-calculator-ii/0227-basic-calculator-ii.py
class Solution:
    def calculate(self, s: str) -> int:
        stack=[]
        num=0
        op='+'
        for i in range(len(s)+1):
            ch='+' if i==len(s) else s[i]

            if ch.isdigit():
                num=num*10+int(ch)
            if i==len(s) or ch in '+-*/':
                if op=='+':
                    stack.append(num)
                elif op=='-':
                    stack.append(-num)
                elif op=='*':
                    stack.append(stack.pop()*num)

                elif op=='/':
                    stack.append(int(stack.pop()/num))
                op=ch
                num=0

        return sum(stack)
        